extensions/interactions/EndExploration/directives/oppia-interactive-end-exploration.component.spec.ts

```diff
@@ -0,0 +1,239 @@
+    const requestUrl = '/explorationsummarieshandler/data?' +
+      'stringified_exp_ids=' + encodeURI(JSON.stringify(explorationIds));
+
+    beforeEach(angular.mock.module('oppia', function($provide) {
+      $provide.value(
+        'InteractionAttributesExtractorService',
+        mockInteractionAttributesExtractorService);
+      $provide.value('$attrs', {
+        recommendedExplorationIds: explorationIds
+      });
+    }));
+
+    beforeEach(angular.mock.inject(function($injector, $componentController) {
+      $q = $injector.get('$q');
+      $rootScope = $injector.get('$rootScope');
+      $scope = $rootScope.$new();
+      ContextService = $injector.get('ContextService');
+      ReadOnlyCollectionBackendApiService =
+        $injector.get('ReadOnlyCollectionBackendApiService');
+      $httpBackend = $injector.get('$httpBackend');
+      UrlService = $injector.get('UrlService');
+
+      ctrl = $componentController('oppiaInteractiveEndExploration');
+
+      spyOn(ContextService, 'getPageContext').and
+        .returnValue(pageContextEditor);
+      spyOn(ContextService, 'getEditorTabContext').and.
+        returnValue(editorTabContext);
+      spyOn(ReadOnlyCollectionBackendApiService, 'loadCollectionAsync')
+        .and.callFake(function() {
+          var deferred = $q.defer();
+          deferred.resolve(sampleCollection.collection);
+          return deferred.promise;
+        });
+      spyOn(UrlService, 'getCollectionIdFromExplorationUrl')
+        .and.returnValue('0');
+      spyOn(UrlService, 'isIframed').and.returnValue(true);
+
+      $httpBackend.expectGET(requestUrl).respond(httpResponse);
+      ctrl.$onInit();
+      $httpBackend.flush();
+      $scope.$apply();
+    }));
+
+    it('should initialize ctrl variables', function() {
+      expect(ctrl.isIframed).toBe(true);
+      expect(ctrl.isInEditorPage).toBe(true);
+      expect(ctrl.isInEditorPreviewMode).toBe(true);
+      expect(ctrl.isInEditorMainTab).toBe(false);
+      expect(ctrl.collectionId).toBe('0');
+      expect(ctrl.getCollectionTitle()).toBe('Collection Under Test');
+      expect(ctrl.errorMessage).toBe('');
+    });
+
+    it('should not display error message', function() {
+      expect(ctrl.isInEditorPage).toBe(true);
+      expect(ctrl.isInEditorPreviewMode).toBe(true);
+      expect(ctrl.errorMessage).toBe('');
+    });
+  });
+
+  describe('Invalid exploration Id provided', function() {
+    const explorationIds = ['0', '1'];
+    const requestUrl = '/explorationsummarieshandler/data?' +
+      'stringified_exp_ids=' + encodeURI(JSON.stringify(explorationIds));
+
+    beforeEach(angular.mock.module('oppia', function($provide) {
+      $provide.value(
+        'InteractionAttributesExtractorService',
+        mockInteractionAttributesExtractorService);
+      $provide.value('$attrs', {
+        recommendedExplorationIds: explorationIds
+      });
+    }));
+
+    beforeEach(angular.mock.inject(function($injector, $componentController) {
```

Comment:
Can you move the common initialization steps from both the describe blocks outside (into the outer describe block)? To prevent code duplication.

Reply:
Placing `beforeEach(angular.mock.inject(function($injector, $componentController) {` before `beforeEach(angular.mock.module('oppia', function($provide) {` will cause errors.

and line 130 cannot be made common since the exploration ids are different for each `describe`

Reply:
```
[$injector:unpr] Unknown provider: $attrsProvider <- $attrs
```